restli-common/src/main/java/com/linkedin/restli/internal/common/URIMaskUtil.java

```diff
@@ -121,161 +121,161 @@ public static MaskTree decodeMaskUriFormat(StringBuilder toparse) throws Illegal
+                      break;
+                    }
+                    else if (ch == ')')
+                    {
+                      state = ParseState.ASCEND;
+                      nextToken = subPosition;
+                      break;
+                    }
+                    else if (Character.isDigit(ch))
+                    {
+                      // Accumulate the value as it's being read in order
+                      maskValue = maskValue * 10 + Character.getNumericValue(ch);
+                    }
+                    else
+                    {
+                      throw new IllegalMaskException("Malformed mask syntax: unexpected range value");
+                    }
+                    subPosition++;
+                  }
                 }
                 else
                 {
-                  throw new IllegalMaskException("Malformed mask syntax: unexpected range value");
-                }
-                ii++;
-              }
+                  if (toparse.charAt(subPosition + 1) != '(')
+                  {
+                    throw new IllegalMaskException("Malformed mask syntax: expected '(' token");
+                  }
 
-              // Set the mask value to the range value specified for the parameter
-              maskValue = Integer.valueOf(rangeValue.toString());
+                  state = ParseState.DESCEND;
+                  nextToken = subPosition;
+                }
+                break;
+              case ')':
+                state = ParseState.ASCEND;
+                nextToken = subPosition;
+                break;
+              default:
+                field.append(c);
+                break;
             }
-            else
+            if (nextToken != -1)
             {
-              if (toparse.charAt(ii + 1) != '(')
-              {
-                throw new IllegalMaskException("Malformed mask syntax: expected '(' token");
-              }
-
-              state = ParseState.DESCEND;
-              nextToken = ii;
+              break;
+            }
+          }
+          if (toparse.length() - position != field.length())
+          {
+            if (nextToken == -1)
+            {
+              throw new IllegalMaskException("Malformed mask syntax: expected closing token");
+            }
+            position = nextToken;
+          }
+          else
+          {
+            position = toparse.length();
+          }
+          if (state == ParseState.DESCEND)
+          {
+            if (field.length() == 0)
+            {
+              throw new IllegalMaskException("Malformed mask syntax: empty parent field name");
             }
-            break;
-          case ')':
-            state = ParseState.ASCEND;
-            nextToken = ii;
-            break;
-          default:
-            field.append(c);
-            break;
+            DataMap subTree = new DataMap();
+            stack.peekLast().put(field.toString().trim(), subTree);
```

Comment:
trim() makes a copy. you can avoid that by checking if a char is whitespace before adding to field.